_posts/2022-12-21-easily-verify-your-rust-in-ci-with-kani.md

```diff
@@ -0,0 +1,80 @@
+---
+layout: post
+title: "Easily verify your Rust in CI with Kani and Github Actions"
+---
+
+Formal verification tools, such as Kani, are valuable because they provide a way to definitively check, using mathematical techniques, whether a property of your code is true under all circumstances.
+To put it colloquially: testing is great for catching the **first** bug, formal verification is great for knowing you’ve caught the **last** bug.
+To learn more about Kani, and how you can use it to prevent bugs in your Rust code, you can [read the Kani documentation](https://model-checking.github.io/kani/), or [consult the real-world examples on our blog](../../..).
+
+Code, however, is seldom static.
```

Comment:
> by ensuring that tests are re-run on all code-changes, CI

> provide facilities to automatically incorporate CI...run checks, ranging from tests, to linters...

These sentences seem to imply that CI is a service/tool for running tests upon code changes. [Continuous Integration](https://en.wikipedia.org/wiki/Continuous_integration) is much broader than that, it is a *process* of continuously integrating code into a central repo, and having an automated testing setup is just one small aspect of it.

Mentioning this because "Continuous Verification" below seems to be building on that analogy, and I'm not sure that it works. I think that running the proofs is part of CI, just like running the tests is.
```suggestion
Code, however, is seldom static. As code evolves, how can we ensure that it remains correct; i.e. how do we catch the *next* bug? Many codebases already contain test suites and other checks that can run automatically. Developers can use a platform like [GitHub Actions](https://github.com/features/actions) to run these tests upon every code change, as part of continuous integration, and even [ensure that the tests pass before the code can be merged](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging). This assures developers that a code change does not introduce a bug for all the inputs that the test suite exercises.

For developers looking for a higher assurance of software quality, Kani proofs can be validated automatically as part of continuous integration. To this end, we are excited to announce the[Kani Rust Verifier Action](https://github.com/marketplace/actions/kani-rust-verifier) on the GitHub Marketplace.
```